jep/401/README.adoc

```diff
@@ -0,0 +1,245 @@
+| Status
+| Not Submitted :information_source:
+
+| Type
+| :bulb: Standards, Informational, or Process :bulb:
+
+| Created
+| :bulb: Date (YYYY-MM-DD) :bulb:
+
+| BDFL-Delegate
+| TBD
+
+//
+//
+// Uncomment if there is an associated placeholder JIRA issue.
+//| JIRA
+//| :bulb: https://issues.jenkins-ci.org/browse/JENKINS-nnnnn[JENKINS-nnnnn] :bulb:
+//
+//
+// Uncomment if discussion will occur in forum other than jenkinsci-dev@ mailing list.
+//| Discussions-To
+//| :bulb: Link to where discussion and final status announcement will occur :bulb:
+//
+//
+// Uncomment if this JEP depends on one or more other JEPs.
+//| Requires
+//| :bulb: JEP-NUMBER, JEP-NUMBER... :bulb:
+//
+//
+// Uncomment and fill if this JEP is rendered obsolete by a later JEP
+//| Superseded-By
+//| :bulb: JEP-NUMBER :bulb:
+//
+//
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+== Abstract
+
+[TIP]
+====
+Give a short (200 word) description of the technical issue addressed.
+
+* Use present tense - describe what the proposal "does" (as if it were already done), not what it will do.
+* Do not go into technical details and instead put those in the Specification section.
+* Do not talk about history or why this needs to be done. Instead, add the history to the Motivation section.
+====
+The permission types `Jenkins.RUN_SCRIPTS`, `Jenkins.CONFIGURE_UPDATECENTER`, and `Jenkins.UPLOAD_PLUGINS` provide a false sense of security, as a user is able to escalate this limited privilege to `Jenkins.ADMINISTER`.  These permissions are being marked as `@Deprecated` and are now consolidated as `Jenkins.ADMINISTER`.  
+
+A new permission type, `Jenkins.CONFIGURE` (implied by `Jenkins.ADMINISTER`) has been created for actions that allow for the management of Jenkins configuration items that do not require `Jenkins.ADMINISTER`.  This new perimssion is used to create a clear separation between settings that impact the entirety of a Jenkins instance (particularly with regards to security) and those that can be used to manage other aspects of an installation. 
+
+
+
+== Specification
+
+[TIP]
+====
+Provide a detailed specification of what is being proposed.
+Be as technical and detailed as needed to allow new or existing Jenkins developers
+to reasonably understand the scope/impact of an implementation.
+
+* Use present tense - describe what the proposal "does" (as if it were already done), not what it will do.
+* Do not discuss alternative designs that were rejected - those belong in the Reasoning section.
+* Avoid in-depth discussion or justification of design choices - that belongs in the Reasoning section.
+====
+
+[]
+====
+Angelique - need to add additional info about Jenkins.READ?
+====
+
+The following permission types are now marked as `@Deprecated` and `@Restricted(NoExternalUse.class)`:
```

Comment:
This is not what `@Restricted` is for, at all. If you want to deprecate something, mark it `@Deprecated`, add reasoning in the Javadoc `@deprecated` tag, clean up all usages you can find, and stop there.

Reply:
I'm on the fence with this.
in one way making it Restricted means that plugin authors have to adapt when using new cores and can not miss that a general administrator would be able to perform the action even though they are checking `RUN_SCRIPTS`.

it won't break anything existing.

At the same time you can bypass the annotations so having them deprecated also helps.  and when upgrading Jenkins baseline you have so much deprecated code in your plugin anyway that you would probably miss it.

Could / should we change the permission so that they are actually just the ADMINISTER permission?

```java
/**
 * @deprecated - use #ADMINSITER
 *
*/
@Deprecated
Permission RUN_SCRIPTS = Jenkins.ADMINISTER

```

Reply:
`@Restricted` is the way to hide public members from the public API. This is what we want to do here. We want plugin developers to progressively stop using those permissions (as new plugin releases come), and specially **force new code to not use it**.

However, I agree that it is not semantically correct, as any `@Restricted` API should be removable or incompatibly evolved, and that's not true in this case. 

I wonder if there should be a new annotation to achieve this, something like a hard deprecation to indicate that the API is deprecated and will be marked as `@Restricted` at some point, and finally removed. But this is probably another discussion to have.

Reply:
Perhaps what you desire is `@RestrictedSince` like in https://github.com/jenkinsci/jenkins/blob/b26fff4f37bd6ef3e7788219af05ad03152c1f3d/core/src/main/java/hudson/org/apache/tools/tar/TarInputStream.java#L43-L47.

In addition to the `@Restricted`, you know that the code *was* public somewhere in the past and could break compatibility if removed.